linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/IncOptimizer.scala

```diff
@@ -79,50 +80,72 @@ final class IncOptimizer private[optimizer] (config: CommonPhaseConfig, collOps:
     }
 
-    val newLinkedClasses = for (linkedClass <- unit.classDefs) yield {
-      val className = linkedClass.className
-      val interface = getInterface(className)
+    val newLinkedClasses = unit.classDefs.map(optimizedLinkedClass(_))
+    val newTopLevelExports = unit.topLevelExports.map(optimizedTopLevelExport(_))
 
-      val publicContainer = classes.get(className).getOrElse {
-        /* For interfaces, we need to look at default methods.
-         * For other kinds of classes, the public namespace is necessarily
-         * empty.
-         */
-        val container = interface.staticLike(MemberNamespace.Public)
-        assert(
-            linkedClass.kind == ClassKind.Interface || container.methods.isEmpty,
-            linkedClass.className -> linkedClass.kind)
-        container
-      }
+    new LinkingUnit(unit.coreSpec, newLinkedClasses, newTopLevelExports,
+        unit.moduleInitializers)
+  }
 
-      val newMethods = for (m <- linkedClass.methods) yield {
-        val namespace = m.flags.namespace
-        val container =
-          if (namespace == MemberNamespace.Public) publicContainer
-          else interface.staticLike(namespace)
-        container.methods(m.methodName).optimizedDef
-      }
+  private def optimizedLinkedClass(linkedClass: LinkedClass): LinkedClass = {
+    val className = linkedClass.className
+    val interface = getInterface(className)
 
-      linkedClass.optimized(newMethods, interface.optimizedExportedMembers(),
-          interface.optimizedJSConstructorDef())
+    val publicContainer = classes.get(className).getOrElse {
+      /* For interfaces, we need to look at default methods.
+       * For other kinds of classes, the public namespace is necessarily
+       * empty.
+       */
+      val container = interface.staticLike(MemberNamespace.Public)
+      assert(
+          linkedClass.kind == ClassKind.Interface || container.methods.isEmpty,
+          linkedClass.className -> linkedClass.kind)
+      container
     }
 
-    new LinkingUnit(unit.coreSpec, newLinkedClasses, unit.topLevelExports,
-        unit.moduleInitializers)
+    val newMethods = for (m <- linkedClass.methods) yield {
+      val namespace = m.flags.namespace
+      val container =
+        if (namespace == MemberNamespace.Public) publicContainer
+        else interface.staticLike(namespace)
+      container.methods(m.methodName).optimizedDef
+    }
+
+    linkedClass.optimized(newMethods, interface.optimizedExportedMembers(),
+        interface.optimizedJSConstructorDef())
+  }
+
+  private def optimizedTopLevelExport(linked: LinkedTopLevelExport): LinkedTopLevelExport = {
+    linked.tree match {
+      case method: TopLevelMethodExportDef =>
+        val newMethod =
+          topLevelExports.optimizedMethod(method.moduleID, method.topLevelExportName)
+
+        new LinkedTopLevelExport(linked.owningClass, newMethod,
+            linked.staticDependencies, linked.externalDependencies)
```

Comment:
The tests fail with module splitting, because this line is wrong: The static dependencies of an optimized method are not necessarily the same as the one of an un-optimized method.

In fact: This surfaces a latent issue we've had since module splitting was introduced: The optimizer does not return correct LinkedClasses anymore: The optimizer can (and does) change the real dependencies.

This issue has not been visible before, because the Refiner always recalculated all linking related infos.

Reply:
Updated. PTAL. I'm toying with a refactor that might allow the optimizer to not have to lie.

Reply:
(for a separate PR that is, IMO this can go in like this nevertheless).